document/core/valid/instructions.rst

```diff
@@ -232,6 +232,439 @@ Reference Instructions
+   }
+
+
+.. _valid-array.get:
+.. _valid-array.get_u:
+.. _valid-array.get_s:
+
+:math:`\ARRAYGET\K{\_}\sx^?~x`
+..............................
+
+* The :ref:`defined type <syntax-deftype>` :math:`C.\CTYPES[x]` must exist.
+
+* The :ref:`expansion <aux-expand>` of :math:`C.\CTYPES[x]` must be an :ref:`array type <syntax-arraytype>` :math:`\TARRAY~\fieldtype`.
+
+* Let the :ref:`field type <syntax-fieldtype>` :math:`\mut~\storagetype` be :math:`\fieldtype`.
+
+* Let :math:`t` be the :ref:`value type <syntax-valtype>` :math:`\unpacktype(\storagetype)`.
+
+* The extension :math:`\sx` must be present of and only if :math:`t` is different from :math:`\storagetype`.
+
+* Then the instruction is valid with type :math:`[(\REF~\NULL~x)~\I32] \to [t]`.
+
+.. math::
+   \frac{
+     \expand(C.\CTYPES[x]) = \TARRAY~(\mut~\X{st})
+     \qquad
+     \sx = \epsilon \Leftrightarrow \X{st} = \unpacktype(\X{st})
+   }{
+     C \vdashinstr \ARRAYGET\K{\_}\sx^?~x : [(\REF~\NULL~x)~\I32] \to [\unpacktype(\X{st})]
+   }
+
+.. _valid-array.set:
+
+:math:`\ARRAYSET~x~i`
+.....................
+
+* The :ref:`defined type <syntax-deftype>` :math:`C.\CTYPES[x]` must exist.
+
+* The :ref:`expansion <aux-expand>` of :math:`C.\CTYPES[x]` must be an :ref:`array type <syntax-arraytype>` :math:`\TARRAY~\fieldtype`.
+
+* Let the :ref:`field type <syntax-fieldtype>` :math:`\mut~\storagetype` be :math:`\fieldtype`.
+
+* The prefix :math:`\mut` must be :math:`\MVAR`.
+
+* Let :math:`t` be the :ref:`value type <syntax-valtype>` :math:`\unpacktype(\storagetype)`.
+
+* Then the instruction is valid with type :math:`[(\REF~\NULL~x)~\I32~t] \to []`.
+
+.. math::
+   \frac{
+     \expand(C.\CTYPES[x]) = \TSTRUCT~(\MVAR~\X{st})
+   }{
+     C \vdashinstr \ARRAYSET~x : [(\REF~\NULL~x)~\I32~\unpacktype(\X{st})] \to []
+   }
+
+.. _valid-array.len:
+
+:math:`\ARRAYLEN`
+.................
+
+* The the instruction is valid with type :math:`[(\REF~\NULL~\ARRAY)] \to [\I32]`.
+
+.. math::
+   \frac{
+   }{
+     C \vdashinstr \ARRAYLEN : [(\REF~\NULL~\ARRAY)] \to [\I32]
+   }
+
+
+
+.. index:: scalar reference
+
+Scalar Reference Instructions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _valid-i31.new:
+
+:math:`\I31NEW`
+...............
+
+* The instruction is valid with type :math:`[\I32] \to [(\REF~\I31)]`.
+
+.. math::
+   \frac{
+   }{
+     C \vdashinstr \I31NEW : [\I32] \to [(\REF~\I31)]
+   }
+
+.. _valid-i31.get_sx:
+
+:math:`\I31GET\K{\_}\sx`
+........................
+
+* The instruction is valid with type :math:`[(\REF~\I31)] \to [\I32]`.
+
+.. math::
+   \frac{
+   }{
+     C \vdashinstr \I31GET\K{\_}\sx : [(\REF~\I31)] \to [\I32]
+   }
+
+
+.. index:: external reference
+
+External Reference Instructions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _valid-extern.internalize:
+
+:math:`\EXTERNINTERNALIZE`
+..........................
+
+* The instruction is valid with type :math:`[(\REF~\NULL_1^?~\EXTERN)] \to [(\REF~\NULL_2^?~\ANY)]` for any :math:`\NULL_1^?` that equals :math:`\NULL_2^?`.
```

Comment:
Does this kind of nullability constraint run afoul of our new principal typing rules?